src_firmware/src/driver.c

```diff
@@ -4,296 +4,485 @@
-		ret = gpio_pin_set_dt(&set_dir_p1, 0);
+		ret = gpio_pin_set_dt(&(drv_chnls[chnl].set_dir_pins[P0]), 0);
 
 		if (ret != 0) {
 			return UNABLE_TO_SET_GPIO;
 		}
-		ret = gpio_pin_set_dt(&set_dir_p2, 0);
+		ret = gpio_pin_set_dt(&(drv_chnls[chnl].set_dir_pins[P1]), 0);
 		if (ret != 0) {
 			return UNABLE_TO_SET_GPIO;
 		}
-		is_motor_on = false;
+		drv_chnls[chnl].is_motor_on = false;
 		return SUCCESS;
 	}
 
 	return NOT_INITIALISED;
 }
+bool get_motor_off_on(enum ChannelNumber chnl)
+{
+	return drv_chnls[chnl].is_motor_on;
+}
+
+// Speed control functions
+static int speed_pwm_set(uint32_t value)
+{
+	int ret;
+
+	if (!drv_initialised) {
+		return NOT_INITIALISED;
+	}
+
+	if (value > CONFIG_SPEED_MAX_MRPM) {
+		return DESIRED_VALUE_TO_HIGH;
+	}
+
+	if (drv_chnls[CH0].target_speed_mrpm < CONFIG_SPEED_MAX_MRPM/10) {
+		value = 0;
+		drv_chnls[CH0].speed_control = 0;
+	}
+
+	uint64_t w_1 = drv_chnls[CH0].pwm_motor_driver.period * (uint64_t)value;
+	uint32_t w = value != 0 ? (uint32_t)(w_1/CONFIG_SPEED_MAX_MRPM) : 0;
+
+	ret = pwm_set_pulse_dt(&(drv_chnls[CH0].pwm_motor_driver), w);
+	if (ret != 0) {
+		return UNABLE_TO_SET_PWM_CHNL1;
+	}
+
+	return SUCCESS;
+}
+int target_speed_set(uint32_t value, enum ChannelNumber chnl)
+{
+	if (control_mode != SPEED) {
+		return UNSUPPORTED_FUNCTION_IN_CURRENT_MODE;
+	}
+	drv_chnls[chnl].target_speed_mrpm = value;
+	drv_chnls[chnl].count_cycles = 0;
+	drv_chnls[chnl].old_count_cycles = 0;
+	return SUCCESS;
+}
+int speed_get(enum ChannelNumber chnl, uint32_t *value)
+{
+	if (drv_initialised) {
+		*value = drv_chnls[chnl].actual_mrpm; // TODO - get speed from encoder
+		return SUCCESS;
+	}
 
+	return NOT_INITIALISED;
+}
 uint32_t speed_target_get(void)
 {
-	return target_speed_mrpm;
+	return drv_chnls[CH0].target_speed_mrpm;
+}
+uint32_t get_current_max_speed(void)
+{
+	return CONFIG_SPEED_MAX_MRPM;
 }
 
-#if defined(CONFIG_BOARD_NRF52840DONGLE_NRF52840)
-void enter_boot(void)
+// position control functions
+int target_position_set(uint32_t new_target_position, enum ChannelNumber chnl)
 {
-	gpio_pin_configure_dt(&out_boot, GPIO_OUTPUT);
+	if (!drv_initialised) {
+		return NOT_INITIALISED;
+	}
+
+	if (control_mode != POSITION) {
+		return UNSUPPORTED_FUNCTION_IN_CURRENT_MODE;
+	}
+
+	drv_chnls[chnl].target_position = new_target_position;
+	return SUCCESS;
 }
-#endif
+int position_get(uint32_t *value, enum ChannelNumber chnl)
+{
+	if (!drv_initialised) {
+		return NOT_INITIALISED;
+	}
+
+	*value = drv_chnls[chnl].curr_pos;
+	return SUCCESS;
+}
+
+// Mode is channel-independent
+int mode_set(enum ControlModes new_mode)
+{
+	if (!drv_initialised) {
+		return NOT_INITIALISED;
+	}
 
-bool get_motor_off_on(void)
+	control_mode = new_mode;
+	return SUCCESS;
+}
+int mode_get(enum ControlModes *value)
 {
-	return is_motor_on;
+	if (!drv_initialised) {
+		return NOT_INITIALISED;
+	}
+
+	*value = control_mode;
+	return SUCCESS;
 }
+int get_control_mode_from_string(char *str_control_mode, enum ControlModes *ret_value)
```

Comment:
Why do we use strings if we have a well working shell with arguments configurations?

Reply:
We didn't have well working arguments back then, added TODO for whole shell.c refactor as this isn't the only case where we could use arguments